### DISCUSSION TITLE
Server side application rendering

### DISCUSSION THREAD

**Discussion by saeedtahmtan** [CuarzoSoftware/Louvre#81] [Ideas]
Hello!

since i meet the great server side decoration of louvre-view and smoothness of it. i have new idea to render an application entirely in server side using read a html syntax like and render it.

i just have some question :
   is it possible with current server side decorations (
       button (click to run a linux command),
       text,
       img,
   ) to make somthing like 
![BluetoothMacSettings](https://github.com/user-attachments/assets/297b9a2b-9194-4e03-99df-622e672bd68f)

   is it a bad idea ?
   does it impact performance ?
   any suggestion for implementation ?

**Comment by ehopperdietzel** [comment]
Hi, I think it would be better to make it a separate app, as rendering HTML could slow down the compositor. I assume you're asking this because you simply want to create/draw UI elements from the compositor.

Louvre uses OpenGL for rendering, so you have a lot of flexibility. If you need to use a CPU rasterizer toolkit (like an HTML rendering tool), you can create an LTexture from the rendered HTML pixel data and update it as necessary to later render it on screen. I do this for rendering text in the examples.

Then to simplify rendering, instead of creating your own GLES shaders or using LPainter, you could use LScene and the various LView subtypes. For instance, LTextureView lets you specify a source texture rectangle, a destination rectangle on the screen, and any custom transformations needed (like rotation or flipping), clipping, etc.

You can think of LView as the base class for creating UI components.

**Reply by saeedtahmtan** [reply]
not necessarily HTML :) it was just an example for have an interpreter.

I can use opcodes to doing that purpose. having separate process not good for me because i want make new way to render native apps faster than any other apps.

cache apps and not have same design pattern are my major problem for make applications.

thank you for your advises they are so helpful to me.

**Reply by ehopperdietzel** [reply]
Ok, well, in any case, client-side rendering is not necessarily slower. If you create a GPU-accelerated app (e.g., with EGL or using GBM and dma buffers and APIs like OpenGL/Vulkan), you can pass GPU buffer handles directly to the compositor without involving any copying, making it super efficient. Also, by distributing the rendering load across multiple processes, you reduce the load on the compositor and can therefore achieve superior performance. In some cases, like fullscreen windows, the compositor can directly display that buffer on the screen without having to render it using OpenGL, which significantly reduces latency and GPU consumption.

Now, if you don't want to use Wayland, you could use SRM to render content directly on the screen (as if it were a compositor) and implement your own display server protocol.

Here's an SRM example which uses Skia:

https://github.com/ehopperdietzel/srm-skia

**Reply by dblsaiko** [reply]
> render native apps faster than any other apps

The only way that would matter is if you were on an extremely resource constrained system (embedded). Are you?

Otherwise you'd put a lot of effort into something for absolutely no benefit, because Wayland protocol and a separate process is not going to be your bottleneck. Just write a normal application.

EDIT: Actually, guaranteed your "HTML" or whatever renderer is going to be so much more expensive than what you would save by building your application in the compositor.

**Reply by saeedtahmtan** [reply]
I was planning to create system tools like (calculator, settings, file manager, etc.) with Electron! But as you know it is not a good option.
OpenGL, Vulkan and all the other major graphics libraries are great options and I definitely agree with that, but they are too complex for me and require a lot of time.

I will take your advice and go with it to create those applications with the OpenGL.

**Reply by dblsaiko** [reply]
That was not what my point was. My point was, don't do this in the compositor.

**Reply by saeedtahmtan** [reply]
Anyway, I agree and don't do it in compositor. It was just an idea. thank you very much
